### ISSUE TITLE
oss_ddp will assert in forward when running in eval mode

### ISSUE THREAD

**Issue by msbaines** [facebookresearch/fairscale#27]
## 🐛 Bug

In training, we only forward which will cause the need_reduction check to fail. This can be fixed setting need_reduction in a backward hook instead of forward.

## To Reproduce

1.
2.
3.

## Environment

Please copy and paste the output from the
environment collection script from PyTorch
(or fill out the checklist below manually).

You can run the script with:
```
# For security purposes, please check the contents of collect_env.py before running it.
python -m torch.utils.collect_env
```

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

**Comment by min-xu-ai** [comment]
Thanks for catching. Just to make sure, is this blocking anything? If not, I will do it later and add a unit test for it too.

**Comment by myleott** [comment]
It's not highpri, but this blocks replacing fairseq's LegacyDistributedDataParallel with the one in fairscale. Here is the fix we did in fairseq: https://github.com/pytorch/fairseq/commit/631023200a8e0c8531ccdb9e5d7f54411149efdf

**Comment by min-xu-ai** [comment]
Thanks, Myle! Will try to get this fix this week.
